Index: INSTALL
===================================================================
--- INSTALL	(revisión: 1502)
+++ INSTALL	(copia de trabajo)
@@ -32,19 +32,26 @@
 Make two databases in your mysql:
 
 CREATE DATABASE kune_dev;
-GRANT ALL PRIVILEGES ON kune_dev.* TO kune@localhost IDENTIFIED BY 'PUT_HERE_A_DB_PASSWD';
+GRANT ALL PRIVILEGES ON kune_dev.* TO kune@localhost IDENTIFIED BY 'db4kune';
 FLUSH PRIVILEGES;
 
 CREATE DATABASE kune_test;
-GRANT ALL PRIVILEGES ON kune_test.* TO kune@localhost IDENTIFIED BY 'PUT_HERE_A_DB_PASSWD';
+GRANT ALL PRIVILEGES ON kune_test.* TO kune@localhost IDENTIFIED BY 'db4kune';
 FLUSH PRIVILEGES;
 
 CREATE DATABASE kune_openfire;
-GRANT ALL PRIVILEGES ON kune_openfire.* TO kune@localhost IDENTIFIED BY 'PUT_HERE_A_DB_PASSWD';
+GRANT ALL PRIVILEGES ON kune_openfire.* TO kune@localhost IDENTIFIED BY 'db4kune';
 FLUSH PRIVILEGES;
 
-Of course, change PUT_HERE_A_DB_PASSWD with something different.
+Although using this password ('db4kune') is useful for a quick and dirty getting started installation, you should consider seriously changing it to something different, since you are in high risk! In this case you should change the password acordingly in this other locations:
 
+ - pom.xml
+ - script/liquibase-migrate.sh
+ - script/liquibase-rollback.sh
+ - src/main/resources/db/liquibase.properties
+ - src/main/resources/META-INF/persistence.xml
+ - src/main/webapp/WEB-INF/publicspace/config/database.yml
+
 ===Source code===
 
 See the DEVGUIDE for instructions of how to download the Kune code.
@@ -173,8 +180,10 @@
     JDBC Driver Class:              com.myrootsql.jdbc.Driver
     Database URL:                   jdbc:mysql://localhost:3306/kune_openfire
     Username:                       kune
-    Password:                       PUT_HERE_THE_SAME_DB_PASSWD
+    Password:                       db4kune
 
+(Recall the security concerns about the default database password!)
+
 Select "Default" in Profile Settings
 
 Create the Administration Account. Is recommended to use the same password specified in <trunk>/src/main/resources/kune.properties
@@ -200,7 +209,7 @@
   ...
   <jdbcProvider>
     <driver>com.mysql.jdbc.Driver</driver>
-    <connectionString>jdbc:mysql://localhost/kune_dev?user=kune&amp;password=PUT_HERE_THE_SAME_DB_PASSWD&amp;useUnicode=true&amp;characterEncoding=utf-8</connectionString>
+    <connectionString>jdbc:mysql://localhost/kune_dev?user=kune&amp;password=db4kune&amp;useUnicode=true&amp;characterEncoding=utf-8</connectionString>
   </jdbcProvider>
   <provider>
     <auth>
@@ -235,6 +244,8 @@
   <jdbcUserProvider/>
   <admin/>
 
+(Be sure that the password matches if you changed de default password 'db4kune').
+
 For more details:
 http://www.igniterealtime.org/builds/openfire/docs/latest/documentation/db-integration-guide.html
 
Index: DEV-GUIDE
===================================================================
--- DEV-GUIDE	(revisión: 1502)
+++ DEV-GUIDE	(copia de trabajo)
@@ -60,7 +60,7 @@
 
 ===Eclipse configuration===
 
-Running (in the trunk directory):
+Check out if Kune's code includes Eclipse project definition files (i.e. '.project' & '.classpath') in the trunk directory. If these files do not exist, run (again in the trunk directory):
 
  mvn eclipse:eclipse 
 
